### PR TITLE
Migrate store/actions from editor package to fields package

### PR DESCRIPTION
**Pull request by gigitux**
## What?

This PR continues the work started in [PR #65261](https://github.com/WordPress/gutenberg/pull/65261) by migrating the remaining actions, with the exception of `DuplicateTemplatePart`, which will be addressed in a separate PR.

Changes Introduced

To avoid code duplication and standardize the way actions are written in the codebase, this PR introduces two new functions in the `mutation` folder:
- deletePostWithNotices
- editPostWithNotices

Happy to revisit the naming of these functions! 

These functions handle the updating of entities and manage success and error notices. The logic they encapsulate was previously duplicated across multiple actions, so unifying it into these functions should improve maintainability.​

Currently, there are other entities in this package that update entities without using these two utils. I'm going to refactor them once this PR is merged. 

### Old description

<details>

This PR is a follow-up of https://github.com/WordPress/gutenberg/pull/65261. With this PR, all actions have been migrated except for:

## `deletePostAction`

The blocker for this action is the dispatch of `removeTemplates` action imported by the `@wordpress/editor` package:

https://github.com/WordPress/gutenberg/blob/4356a42b05e569536502f33fd41b69a4e23722f5/packages/editor/src/dataviews/actions/delete-post.tsx#L53-L55

## `duplicateTemplatePart` 

The blocker for this action is `CreateTemplatePartModalContents`:

https://github.com/WordPress/gutenberg/blob/4356a42b05e569536502f33fd41b69a4e23722f5/packages/editor/src/dataviews/actions/duplicate-template-part.tsx#L53-L54

This component is used in the `@wordpress/editor` package by another component.

## `resetPost`

The blocker for this action is the dispatch of `removeTemplates` action imported by the `@wordpress/editor` package:

https://github.com/WordPress/gutenberg/blob/4356a42b05e569536502f33fd41b69a4e23722f5/packages/editor/src/dataviews/actions/reset-post.tsx#L44-L46

<hr />

I'm wondering how we should tackle this. Should we re-implement `removeTemplates` in @wordpress/fields? 🤔 

cc @youknowriad @oandregal 

</details>

## Testing Instructions

Ensure that `Custom Dataviews` is enabled.

1. Visit the Site Editor.
2. Visit the patterns view.
3. Smoke test all the actions available.
4. Visit the templates view.
5. Smoke test all the actions available.
6. Visit the page view.
7. Smoke test all the actions available.

### CODE REVIEW

**Comment by youknowriad** [comment]
> I'm wondering how we should tackle this. Should we re-implement removeTemplates in @wordpress/fields? 🤔

Yes, It can either be moved to the "component" itself or move to core-data. I might prefer in the component directly for now like all the other actions.

File: packages/editor/src/dataviews/actions/delete-post.tsx
Comment (gigitux) [inline packages/editor/src/dataviews/actions/delete-post.tsx:1]: This action has been separated into:

- `delete-template` action
- `delete-pattern` action

File: packages/editor/src/dataviews/actions/reset-post.tsx
Comment (gigitux) [inline packages/editor/src/dataviews/actions/reset-post.tsx:1]: This action has been renamed into `reset-template`.

File: packages/editor/src/dataviews/fields/index.ts
Comment (gigitux) [inline packages/editor/src/dataviews/fields/index.ts:1]: These fields are already available in `@wordpress/fields` package.

File: packages/fields/README.md
Comment (gigitux) [inline packages/fields/README.md:1]: In a dedicated PR, I will improve documentation.

**Comment by gigitux** [comment]
cc @youknowriad @oandregal I'm marking this PR ready for review!

File: packages/fields/src/actions/template/reset-template.tsx
Comment (louwie17) [inline packages/fields/src/actions/template/reset-template.tsx]: These two should be moved up to `WordPress dependencies`

File: packages/fields/src/actions/template/delete-template.tsx
Comment (louwie17) [inline packages/fields/src/actions/template/delete-template.tsx]: This should be moved to to `WordPress dependencies`

File: packages/fields/src/actions/pattern/delete-pattern.tsx
Comment (louwie17) [inline packages/fields/src/actions/pattern/delete-pattern.tsx]: This should be moved up to `WordPress dependencies`

File: packages/fields/src/actions/template/reset-template.tsx
Comment (louwie17) [inline packages/fields/src/actions/template/reset-template.tsx]: Having to copy this from the `core/edit-site` data store makes me wonder if this is something that should be passed in as a config. A `onResetTemplate` function that is optional, but can be passed into the `action`?
Could of-course be a follow up.

File: packages/fields/src/actions/pattern/delete-pattern.tsx
Comment (louwie17) [inline packages/fields/src/actions/pattern/delete-pattern.tsx]: Could the `RenderModal` for this one and `delete-template` be shared?
I realize currently this doesn't have access to the above action, but maybe we can move this into a thin wrapper?
A future idea is to pass the action into the `RenderModal` https://github.com/WordPress/gutenberg/blob/trunk/packages/dataviews/src/components/dataviews-item-actions/index.tsx#L115

File: packages/fields/src/mutation/index.ts
Comment (louwie17) [inline packages/fields/src/mutation/index.ts]: Should move this up.

**Review by louwie17 — COMMENTED** [review]
Overall looks good @gigitux and it tested well. I did notice several places where you still had some `@wordpress/*` imports under the `Internal Dependencies` comment block. I am not sure why the linter didn't catch this, maybe that is missing in the `fields` package still?

I left a couple other thoughts, but those aren't blockers.

File: packages/fields/src/actions/template/reset-template.tsx
Comment (gigitux) [reply packages/fields/src/actions/template/reset-template.tsx]: > Having to copy this from the core/edit-site data store makes me wonder if this is something that should be passed in as a config

From what I understood, it looks like all actions that manipulate WordPress entities should exist in the `@wordpress/core-data` package.

Given that this PR is already big, and we are still evaluating multiple ideas, I followed the [@youknowriad suggestion's ](https://github.com/WordPress/gutenberg/pull/65289#issuecomment-2346618234) to move it into this package.

File: packages/fields/src/actions/pattern/delete-pattern.tsx
Comment (gigitux) [reply packages/fields/src/actions/pattern/delete-pattern.tsx]: >Could the RenderModal for this one and delete-template be shared?

Good point! I would keep them separate and if we will see an overlap with other entities try to create a shared modal.

File: packages/editor/src/dataviews/actions/delete-post.tsx
Comment (youknowriad) [reply packages/editor/src/dataviews/actions/delete-post.tsx:1]: What's the reason for that?

Comment (gigitux) [reply packages/editor/src/dataviews/actions/delete-post.tsx:1]: We decided to separate actions by type, with each type having its own dedicated folder. The only exception is the `common` folder, which contains actions for the `Post` type, as `Post` is a union type of all post types:

https://github.com/WordPress/gutenberg/blob/5f77458225b0ce1d4e5fdb2b1113a374d8bed189/packages/fields/src/types.ts#L65

Since `Template` and `Pattern` are different types, each action should reside in its respective folder.

If you think this approach doesn't make sense, I'm happy to revisit it!

Comment (youknowriad) [reply packages/editor/src/dataviews/actions/delete-post.tsx:1]: As I said in the previous PR, I prefer "flat" structure for this reason. My ideal is that all post types use the same actions, they are just enabled/disabled using "post type supports".

That said for this PR, this could be moved to "common" if it impacts different post types no? instead of duplicating the code.

Comment (gigitux) [reply packages/editor/src/dataviews/actions/delete-post.tsx:1]: > As I said in the previous PR, I prefer "flat" structure for this reason. My ideal is that all post types use the same actions, they are just enabled/disabled using "post type supports".

The main issue is that some actions rely on some property that are not available on other types.

> That said for this PR, this could be moved to "common" if it impacts different post types no? instead of duplicating the code.

It makes sense 👍

Comment (gigitux) [reply packages/editor/src/dataviews/actions/delete-post.tsx:1]: Addressed with 0316536

**Comment by gigitux** [comment]
> Overall looks good @gigitux and it tested well. I did notice several places where you still had some `@wordpress/*` imports under the `Internal Dependencies` comment block. I am not sure why the linter didn't catch this, maybe that is missing in the `fields` package still?
> 
> I left a couple other thoughts, but those aren't blockers.

Fixed with [485237f](https://github.com/WordPress/gutenberg/pull/65289/commits/485237f523c6145b4dd352a876bf81e9b1c91f87)! It is a nice catch, btw! I noticed that other packages suffer the same issue. I'm going to do a quick investigation and open an issue!


EDIT: I opened the issue: https://github.com/WordPress/gutenberg/issues/65440

File: packages/fields/src/actions/template/index.ts
Comment (youknowriad) [inline packages/fields/src/actions/template/index.ts]: This one was also resetPost because it applied to both "template" and "template parts". It's not out of the question that it will apply to "patterns" later. So I'm not sure if having a "template" folder here makes sense, it might be confusing.

Comment (youknowriad) [reply packages/fields/src/actions/template/index.ts]: It's also unclear to me what's the difference between "base-post" and "common" folders.

Comment (gigitux) [reply packages/fields/src/actions/template/index.ts]: `base-post` should contain all the actions that belong to this type:

https://github.com/WordPress/gutenberg/blob/5f77458225b0ce1d4e5fdb2b1113a374d8bed189/packages/editor/src/dataviews/types.ts#L26-L27


`commont` should contain all the actions that belong to the union type:

https://github.com/WordPress/gutenberg/blob/5f77458225b0ce1d4e5fdb2b1113a374d8bed189/packages/editor/src/dataviews/types.ts#L65


It looks like that folder approach is creating more confusion than benefit, I'm going to remove them and create a flat structure 👍 . We can revisit it later!

Comment (gigitux) [reply packages/fields/src/actions/template/index.ts]: Addressed the feedback with 5a4fb59

File: packages/fields/src/module.d.ts
Comment (youknowriad) [inline packages/fields/src/module.d.ts]: What is this about?

Comment (youknowriad) [reply packages/fields/src/module.d.ts]: Why didn't we need something like it before (when the actions were in editor).

Comment (gigitux) [reply packages/fields/src/module.d.ts]: Because we are using `@ts-ignore` to ignore the errors. Both solutions aren't great and given that we are already using `@ts-ignore`, let's remove this file - 0deeed3

File: packages/fields/src/actions/delete-post.tsx
Comment (youknowriad) [inline packages/fields/src/actions/delete-post.tsx:203]: I'm assuming there's no change in logic in the two actions that Github didn't caught as files being moved right?

**Review by youknowriad — APPROVED** [review]

File: packages/fields/src/actions/delete-post.tsx
Comment (gigitux) [reply packages/fields/src/actions/delete-post.tsx:203]: No, I didn't change the logic! Thanks for the review! 🙇

**Comment by oandregal** [comment]
What's the status of this PR? It has an approval, can it be merged, do you need help with that @gigitux ?

File: packages/fields/src/mutation/index.ts
Comment (oandregal) [inline packages/fields/src/mutation/index.ts:1]: Why these actions are not top-level like any other?

**Comment by gigitux** [comment]
> What's the status of this PR? It has an approval, can it be merged, do you need help with that @gigitux ?

I can't merge it! I was thinking that more feedback was necessary before the merge! On my side, would be great if you merge it, so I can unblock https://github.com/WordPress/gutenberg/pull/65390

File: packages/fields/src/mutation/index.ts
Comment (gigitux) [reply packages/fields/src/mutation/index.ts:1]: The idea is that these functions are utils and not “proper actions”

**Comment by oandregal** [comment]
> I can't merge it! I was thinking that more feedback was necessary before the merge! On my side, would be great if you merge it, so I can unblock https://github.com/WordPress/gutenberg/pull/65390

Oh, I see. I will merge then to unblock further work. We can still iterate on this all.

**Comment by oandregal** [comment]
btw, there's some automation to generate the changelog for each Gutenberg (plugin) release. One thing we do is to take the PR names to generate the changelog, for example: https://make.wordpress.org/core/2024/09/12/what-is-new-in-gutenberg-19-2-11-september/ The better names they have, the less work the release coordinator has to do :) I'd suggest would be avoiding formats like `[ XXXX ] YYY` in favor of a more descriptive name, when possible.

File: packages/fields/src/mutation/index.ts
Comment (oandregal) [reply packages/fields/src/mutation/index.ts:1]: Why is the folder called `mutation` and not `utils`?

Comment (gigitux) [reply packages/fields/src/mutation/index.ts:1]: I created a dedicated folder to highlight that these utils “mutate” an entity. I don't have a strong preference, I'm happy to put this file at top level and rename it to `utils.ts`

**Comment by gigitux** [comment]
> btw, there's some automation to generate the changelog for each Gutenberg (plugin) release. One thing we do is to take the PR names to generate the changelog, for example: https://make.wordpress.org/core/2024/09/12/what-is-new-in-gutenberg-19-2-11-september/ The better names they have, the less work the release coordinator has to do :) I'd suggest would be avoiding formats like `[ XXXX ] YYY` in favor of a more descriptive name, when possible.

Thanks for sharing it! 🙇

File: packages/fields/src/mutation/index.ts
Comment (oandregal) [reply packages/fields/src/mutation/index.ts:1]: Looking at the code:

- `editPostWithNotices` is not used anywhere. Can we removed it?
- `deletePostWithNotices` (and the `NoticeSettings` type) is just used by the `delete-post` action. Can we inline it in that file instead?

Comment (gigitux) [reply packages/fields/src/mutation/index.ts:1]: True! To not make this PR too big, I didn't "migrate" existent actions to these two utils, but `deletePostWithNotices` could be used:
- `permanently-delete-post`: 
https://github.com/WordPress/gutenberg/blob/400a67d954eed259514c73d69225e34b08093ef0/packages/fields/src/actions/permanently-delete-post.tsx#L16
- `trash-post`:
https://github.com/WordPress/gutenberg/blob/400a67d954eed259514c73d69225e34b08093ef0/packages/fields/src/actions/restore-post.tsx#L16

Instead `editPostWithNotices` could be used with:
- `restore-post`:
https://github.com/WordPress/gutenberg/blob/400a67d954eed259514c73d69225e34b08093ef0/packages/fields/src/actions/restore-post.tsx#L16
- `reset-post`:
https://github.com/WordPress/gutenberg/blob/400a67d954eed259514c73d69225e34b08093ef0/packages/fields/src/actions/reset-post.tsx#L179

**Pull request merged**
